Clang 8.01 warnings for modcc

Compiling with clang++ 8.0.1 gives a bunch of warnings from modcc code, some of which may well be justified.

Investigate and fix.

```
In file included from /home/bcumming/software/github/arbor/modcc/io/prefixbuf.cpp:6:
/home/bcumming/software/github/arbor/modcc/./io/prefixbuf.hpp:124:5: warning: explicitly defaulted move
      constructor is implicitly deleted [-Wdefaulted-function-deleted]
    pfxstringstream(pfxstringstream&&) = default;
    ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/9.1.0/../../../../include/c++/9.1.0/ostream:58:27: note: move
      constructor of 'pfxstringstream' is implicitly deleted because base class 'basic_ios<char,
      std::char_traits<char> >' has a deleted move constructor
    class basic_ostream : virtual public basic_ios<_CharT, _Traits>
                          ^
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/9.1.0/../../../../include/c++/9.1.0/bits/basic_ios.h:475:7: note: 
      'basic_ios' has been explicitly marked deleted here
      basic_ios(const basic_ios&) = delete;
      ^
1 warning generated.
```

```
/home/bcumming/software/github/arbor/arbor/fvm_layout.cpp:343:23: warning: lambda capture 'D' is not used
      [-Wunused-lambda-capture]
                    [&D](const segment_info& seg_info) { return seg_info.distal_cv; });
                     ~^
```

```
In file included from /home/bcumming/software/github/arbor/arbor/merge_events.cpp:9:
/home/bcumming/software/github/arbor/arbor/io/trace.hpp:59:9: warning: explicitly defaulted move
      constructor is implicitly deleted [-Wdefaulted-function-deleted]
        emit_nl_locked(emit_nl_locked&&) = default;
```

--- a/arbor/fvm_layout.cpp
+++ b/arbor/fvm_layout.cpp
@@ -340,7 +340,7 @@ fvm_discretization fvm_discretize(const std::vector<cable_cell>& cells, const ca
             }
             else {
                 auto opt_index = util::binary_search_index(D.segments, seg_info.parent_cv,
-                    [&D](const segment_info& seg_info) { return seg_info.distal_cv; });
+                    [](const segment_info& seg_info) { return seg_info.distal_cv; });
 
                 if (!opt_index) {
                     throw arbor_internal_error("fvm_layout: could not find parent segment");

--- a/arbor/include/arbor/morph/morphology.hpp
+++ b/arbor/include/arbor/morph/morphology.hpp
@@ -10,7 +10,7 @@
 
 namespace arb {
 
-class morphology_impl;
+struct morphology_impl;
 
 using mindex_range = std::pair<const msize_t*, const msize_t*>;
 

--- a/arbor/io/trace.hpp
+++ b/arbor/io/trace.hpp
@@ -56,8 +56,6 @@ namespace impl {
             lock_(this->guard())
         {}
 
-        emit_nl_locked(emit_nl_locked&&) = default;
-
         ~emit_nl_locked() {
             if (rdbuf()) {
                 (*this) << std::endl;

--- a/arbor/mechcat.cpp
+++ b/arbor/mechcat.cpp
@@ -229,7 +229,7 @@ struct catalogue_state {
         const std::string* base = &name;
 
         if (!defined(name)) {
-            if (implicit_deriv = derive(name)) {
+            if ((implicit_deriv = derive(name))) {
                 base = &implicit_deriv.first().parent;
             }
             else {

--- a/modcc/io/prefixbuf.hpp
+++ b/modcc/io/prefixbuf.hpp
@@ -33,10 +33,16 @@ public:
     explicit prefixbuf(std::streambuf* inner, bool prefix_empty_lines=false):
         inner_(inner), prefix_empty_lines_(prefix_empty_lines) {}
 
-    prefixbuf(prefixbuf&&) = default;
+    prefixbuf(prefixbuf&& other): std::streambuf(other) {
+        prefix = std::move(other.prefix);
+        inner_ = other.inner_;
+        prefix_empty_lines_ = other.prefix_empty_lines_;
+        bol_ = other.bol_;
+    }
+
     prefixbuf(const prefixbuf&) = delete;
 
-    prefixbuf& operator=(prefixbuf&&) = default;
+    prefixbuf& operator=(prefixbuf&&) = delete;
     prefixbuf& operator=(const prefixbuf&) = delete;
 
     std::streambuf* inner() { return inner_; }
@@ -109,20 +115,27 @@ inline indent_manip settab(unsigned w) {
     return indent_manip{indent_manip::settab, w};
 }
 
-// Wrap an stringbuf with a prefixbuf, and present as a stream.
+// Wrap a stringbuf with a prefixbuf, and present as a stream.
 // Acts very much like a `std::ostringstream`, but with prefix
 // and indent functionality.
 
 class pfxstringstream: public std::ostream {
 public:
     pfxstringstream():
-        std::ostream(&pbuf_),
+        std::ostream(nullptr),
         sbuf_(std::ios_base::out),
         pbuf_(&sbuf_)
-    {}
+    {
+        std::ostream::rdbuf(&pbuf_);
+    }
 
-    pfxstringstream(pfxstringstream&&) = default;
-    pfxstringstream& operator=(pfxstringstream&&) = default;
+    pfxstringstream(pfxstringstream&& other):
+        std::ostream(std::move(other)),
+        sbuf_(std::move(other.sbuf_)),
+        pbuf_(std::move(other.pbuf_))
+    {
+        std::ostream::rdbuf(&pbuf_);
+    }
 
     std::string str() const { return sbuf_.str(); }
     void str(const std::string& s) { sbuf_.str(s); }

--- a/modcc/solvers.cpp
+++ b/modcc/solvers.cpp
@@ -1,5 +1,6 @@
 #include <map>
 #include <set>
+#include <stdexcept>
 #include <string>
 #include <vector>
 
@@ -392,12 +393,16 @@ void SparseSolverVisitor::finalize() {
     for (unsigned i = 0; i<A_.nrow(); ++i) {
         const symge::sym_row& row = A_[i];
         unsigned rhs_col = A_.augcol();
-        unsigned lhs_col;
-        for (unsigned r = 0; r < A_.nrow(); r++) {
+        unsigned lhs_col = -1;
+        for (unsigned r = 0; r<A_.nrow(); ++r) {
             if (row[r]) {
                 lhs_col = r;
                 break;
             }
+        }
+
+        if (lhs_col==-1) {
+            throw std::logic_error("zero row in sparse solver matrix");
         }
 
         auto expr =
@@ -478,12 +483,16 @@ void LinearSolverVisitor::finalize() {
     for (unsigned i = 0; i < A_.nrow(); ++i) {
         const symge::sym_row& row = A_[i];
         unsigned rhs = A_.augcol();
-        unsigned lhs;
-        for (unsigned r = 0; r < A_.nrow(); r++) {
+        unsigned lhs = -1;
+        for (unsigned r = 0; r < A_.nrow(); ++r) {
             if (row[r]) {
                 lhs = r;
                 break;
             }
+        }
+
+        if (lhs==-1) {
+            throw std::logic_error("zero row in linear solver matrix");
         }
 
         auto expr =

